pkg/logs/tekton_logging.go

```diff
@@ -194,79 +195,98 @@ func getPipelineRunNamesForActivity(pa *v1.PipelineActivity, tektonClient tekton
 
-	return names, nil
+	return runs, nil
 }
 
 // GetRunningBuildLogs obtains the logs of the provided PipelineActivity and streams the running build pods' logs using the provided LogWriter
 func (t TektonLogger) GetRunningBuildLogs(pa *v1.PipelineActivity, buildName string) error {
-	pipelineRunNames, err := getPipelineRunNamesForActivity(pa, t.TektonClient)
-	if err != nil {
-		return errors.Wrapf(err, "failed to get PipelineRun names for activity %s in namespace %s", pa.Name, pa.Namespace)
-	}
-	pipelineRunsLogged := make(map[string]bool)
+	loggedAllRunsForActivity := false
 	foundLogs := false
 
-	for len(pipelineRunNames) > len(pipelineRunsLogged) {
-		for _, prName := range pipelineRunNames {
-			_, runSeen := pipelineRunsLogged[prName]
-			if !runSeen {
-				structure, err := t.JXClient.JenkinsV1().PipelineStructures(pa.Namespace).Get(prName, metav1.GetOptions{})
-				if err != nil {
-					return errors.Wrapf(err, "failed to get pipeline structure for %s in namespace %s", prName, pa.Namespace)
+	for !loggedAllRunsForActivity {
+		runsByType, err := getPipelineRunsForActivity(pa, t.TektonClient)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get PipelineRun names for activity %s in namespace %s", pa.Name, pa.Namespace)
+		}
+
+		var runToLog *v1alpha12.PipelineRun
+		metaPr := runsByType[tekton.MetaPipeline.String()]
+		buildPr := runsByType[tekton.BuildPipeline.String()]
+		var metaStatus *knativeapis.Condition
+		if metaPr != nil {
+			metaStatus = metaPr.Status.GetCondition(knativeapis.ConditionSucceeded)
+		}
+		// If we don't have a build run yet...
+		if buildPr == nil {
+			// If we have a metapipeline and either the metapipeline is still running or failed, include the metapipeline
+			if metaStatus != nil && metaStatus.Status != corev1.ConditionTrue {
+				runToLog = metaPr
+				// If the metapipeline failed, the pipeline as a whole is complete.
+				if metaStatus.Status == corev1.ConditionFalse {
+					loggedAllRunsForActivity = true
 				}
+			}
+		} else {
+			// Log the build pipeline
+			runToLog = buildPr
+			loggedAllRunsForActivity = true
+		}
+		if runToLog != nil {
+			structure, err := t.JXClient.JenkinsV1().PipelineStructures(pa.Namespace).Get(runToLog.Name, metav1.GetOptions{})
+			if err != nil {
+				return errors.Wrapf(err, "failed to get pipeline structure for %s in namespace %s", runToLog.Name, pa.Namespace)
+			}
 
-				allStages := structure.GetAllStagesWithSteps()
-				stagesSeen := make(map[string]bool)
+			allStages := structure.GetAllStagesWithSteps()
+			stagesSeen := make(map[string]bool)
 
-				// Repeat until we've seen pods for all stages
-				for len(allStages) > len(stagesSeen) {
-					pods, err := builds.GetPipelineRunPods(t.KubeClient, pa.Namespace, prName)
-					if err != nil {
-						return errors.Wrapf(err, "failed to get pods for pipeline run %s in namespace %s", prName, pa.Namespace)
-					}
+			// Repeat until we've seen pods for all stages
+			for len(allStages) > len(stagesSeen) {
```

Comment:
hmm, that's a fair amount of nesting here. By just looking at the code I have a hard time understanding what's happening here? Any chance this would be split up a bit more?